File: Python/Program/ParkingLot.py
from Car import Car
from Motorcycle import Motorcycle

class ParkingLot:
    __kapasitas = 0
    __KendaraanSaatIni = 0
    __vehicles = []
        
    def __init__(self, kapasitas):
        self.__kapasitas = kapasitas
        ParkingLot.__KendaraanSaatIni=0
        self.__vehicles = []
        
    def park_vehicle(self, vehicle):
        if self.__KendaraanSaatIni < self.__kapasitas:
            self.__vehicles.append(vehicle)
            self.__KendaraanSaatIni +=1
        
    # setter getter
    def setKapasitas(self, kapasitas):
        self.__kapasitas = kapasitas
        
    def getKapasitas(self):
        return self.__kapasitas
    
    def setKendaraanSaatIni(self, KendaraanSaatIni):
        self.__KendaraanSaatIni = KendaraanSaatIni
        
    def getKendaraanSaatIni(self):
        return self.__KendaraanSaatIni
    
    def display_vehicle(self):
        print(">>>> List Kendaraan di Parking Lot :")
        for vehicle in self.__vehicles:
            print("Plat Nomor :", vehicle.getNoPlat())
            print("Merk :", vehicle.getMerk())
            print("Tahun Produksi :", vehicle.getTahunProduksi())
            print("Warna :", vehicle.getWarna())
            if isinstance(vehicle, Car):
                print("Banyak Kursi :", vehicle.getJumlahKursi())
                print("Banyak Pintu :", vehicle.getJumlahPintu())
                print("Banyak Roda :", vehicle.getJumlahRoda())
                print("Tipe :", vehicle.getTipe())
            elif isinstance(vehicle, Motorcycle):
                print("Tipe :", vehicle.getTipe())
                print("Kapasitas Tangki :", vehicle.getKapasitasTangki())
            print()    
        print("Kapasitas Tersedia =", self.__kapasitas-self.__KendaraanSaatIni, "dari", self.__kapasitas)

File: CPP/Program/ParkingLot.cpp
#pragma once

#include <string>
#include <iostream>
#include <vector>
#include "Car.cpp"
#include "Motorcycle.cpp"

using namespace std;

class ParkingLot {
private:
    int kapasitas;
    int KendaraanSaatIni;
    vector<Vehicle*> vehicles;

public:
    ParkingLot(int kapasitas) : kapasitas(kapasitas), KendaraanSaatIni(0) {}

    void park_vehicle(Vehicle* vehicle) {
        if (KendaraanSaatIni < kapasitas) {
            vehicles.push_back(vehicle);
            KendaraanSaatIni++;
        }
    }

    int getKapasitas() {
        return kapasitas;
    }

    int getKendaraanSaatIni() {
        return KendaraanSaatIni;
    }

    void display_vehicle() {
        cout << ">>>> List Kendaraan di Parking Lot :\n";
        for (Vehicle* vehicle : vehicles) {
            cout << "Plat Nomor : " << vehicle->getNoPlat() << "\n";
            cout << "Merk : " << vehicle->getMerk() << "\n";
            cout << "Tahun Produksi : " << vehicle->getTahunProduksi() << "\n";
            cout << "Warna : " << vehicle->getWarna() << "\n";
            if (dynamic_cast<Car*>(vehicle)) {
                Car* car = dynamic_cast<Car*>(vehicle);
                cout << "Banyak Kursi : " << car->getJumlahKursi() << "\n";
                cout << "Banyak Pintu : " << car->getJumlahPintu() << "\n";
                cout << "Banyak Roda : " << car->getJumlahRoda() << "\n";
                cout << "Tipe : " << car->getTipe() << "\n";
            } else if (dynamic_cast<Motorcycle*>(vehicle)) {
                Motorcycle* motor = dynamic_cast<Motorcycle*>(vehicle);
                cout << "Tipe : " << motor->getTipe() << "\n";
                cout << "Kapasitas Tangki : " << motor->getKapasitasTangki() << "\n";
            }
            cout << "\n";
        }
        cout << "Kapasitas Tersedia = " << kapasitas - KendaraanSaatIni << " dari " << kapasitas << "\n";
    }
};